### PR TITLE
Implement scoreboard interface

### PR DESCRIPTION
**Pull request by gcobcqwe**

### CODE REVIEW

File: kofserver/scoreboard.py [hunk 1]
```diff
@@ -57,9 +60,21 @@ def __init__(self):
...
     # In this case, Alice should get 30*5 points, Bob should get 30*5+30*10
     # points, and Eve doesn't get any point.
-    def LogTicks(gameName, playerName, portUptime, portScorePerSec, pidUptime, pidScorePerSec):
-        # TODO
-        pass
+    def LogTicks(self, gameName, playerName, portUptime, portScorePerSec, pidUptime, pidScorePerSec):
+        now = int(time.time())
+        sql = "INSERT INTO score (game_name, player_name, port_uptime, port_score_per_sec, pid_uptime, pid_score_per_sec, create_time) VALUES "
+        for i in range(len(playerName)):
+            sql += "('%s', '%s', %d, %d, %d, %d, %d)" % (gameName, playerName[i], portUptime[i], portScorePerSec[i], pidUptime[i], pidScorePerSec[i], now)
```
Comment (john0312) [inline kofserver/scoreboard.py]: Maybe use executemany for better performance and security (sql injection)?

Comment (john0312) [reply kofserver/scoreboard.py]: https://docs.python.org/2/library/sqlite3.html#sqlite3.Cursor.executemany

File: kofserver/scoreboard.py [hunk 2]
```diff
@@ -20,19 +20,22 @@
...
+        self.CreateActionTable()
 
     # This is called every time the Game logic checks if the players are alive.
     # This method is used to record players who are alive.
     # gameName: A string that is the name of the game that we are recording.
     # playerName: An array of player's name.
     # portUptime: An array of uptime for the player's port.
     # portScorePerSec: An array of scores per second of uptime for the player.
-    # pidUptime: An arrya of uptime for the player's process.
+    # pidUptime: An array of uptime for the player's process.
```
Comment (john0312) [inline kofserver/scoreboard.py:38]: Nice find, thanks~

File: kofserver/database.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+                cursor = self._conn.cursor()
+                return cursor.execute(command)     
+            except sqlite3.Error as e:
+                logging.error(e)
+                self._conn = None
+                self.connect()
+                continue
+
+    def Commit(self):
+        self._conn.commit()
```
Comment (john0312) [inline kofserver/database.py]: Nit/Optional: Please add newline at the end of the file

File: kofserver/scoreboard.py [hunk 4]
```diff
@@ -97,6 +121,71 @@ def LogPlayerAction(gameName, playerName, actionType, actionContent):
...
     # [ { "playerName": "Alice", "portUptime": 0, "portScore": 0,
     #     "pidUptime": 60, "pidScore": 300, "totalScore": 300 } ]
-    def QueryScore(gameName, playerName):
-        # TODO
-        return []
+    def QueryScore(self, gameName, playerName):
+        sql = "SELECT player_name, SUM(port_uptime), SUM(port_score_per_sec), SUM(pid_uptime), SUM(pid_score_per_sec) FROM score %s GROUP BY game_name, player_name;"
+        whereSQL = ""
+        if gameName != "":
+            whereSQL += "WHERE game_name='%s'" % (gameName)
```
Comment (john0312) [inline kofserver/scoreboard.py]: SQL injection is possible here, use "?" instead.
https://docs.python.org/2/library/sqlite3.html
"Instead, use the DB-API’s parameter substitution. Put ? as a placeholder wherever you want to use a value, and then provide a tuple of values as the second argument to the cursor’s execute() method."

File: kofserver/scoreboard.py [hunk 5]
```diff
@@ -97,6 +121,71 @@ def LogPlayerAction(gameName, playerName, actionType, actionContent):
...
+        if playerName != "":
+            whereSQL += "WHERE player_name='%s'" % (playerName) if whereSQL == "" else " AND player_name='%s'" % (playerName)
```
Comment (john0312) [inline kofserver/scoreboard.py]: Same here.

File: kofserver/scoreboard.py [hunk 6]
```diff
@@ -97,6 +121,71 @@ def LogPlayerAction(gameName, playerName, actionType, actionContent):
...
+                content TEXT,
+                create_time INTEGER NOT NULL);"""
+        cursor = self.database.Execute(sql)
+        try:
+            self.database.Commit()
+            results = False if cursor.rowcount == -1 else True
+            cursor.close()
+            return results
+        except:  
+            raise
```
Comment (john0312) [inline kofserver/scoreboard.py]: Nit/Optional: Please add new line at the end of the file.

File: kofserver/database.py (hunk 3 above)
Comment (gcobcqwe) [reply kofserver/database.py]: Done

File: kofserver/scoreboard.py (hunk 1 above)
Comment (gcobcqwe) [reply kofserver/scoreboard.py]: Done

File: kofserver/scoreboard.py (hunk 4 above)
Comment (gcobcqwe) [reply kofserver/scoreboard.py]: Done

File: kofserver/scoreboard.py (hunk 6 above)
Comment (gcobcqwe) [reply kofserver/scoreboard.py]: Done

File: kofserver/scoreboard.py (hunk 5 above)
Comment (gcobcqwe) [reply kofserver/scoreboard.py]: Done

**Comment by john0312** [comment]
LGTM, merged.

**Pull request merged**